src/security/KeyData.cc

```diff
@@ -97,33 +98,36 @@ Security::KeyData::loadX509ChainFromFile()
         return;
     }
 
-#if TLS_CHAIN_NO_SELFSIGNED // ignore self-signed certs in the chain
     if (SelfSigned(*cert)) {
-        debugs(83, DBG_PARSE_NOTE(2), "Certificate is self-signed, will not be chained: " << *cert);
-    } else
-#endif
-    {
+        debugs(83, DBG_PARSE_NOTE(2), "Signing certificate is self-signed: " << *cert);
+        // TODO: Warn if there are other (unusable) certificates present.
+    } else {
         debugs(83, DBG_PARSE_NOTE(3), "Using certificate chain in " << certFile);
         // and add to the chain any other certificate exist in the file
         CertPointer latestCert = cert;
 
+        // XXX: The first ca value is usually not a CA certificate because we
+        // loop from the very first certificate in certFilename, and that
+        // certificate is a copy of the already loaded _signing_ this->cert.
         while (const auto ca = Ssl::ReadOptionalCertificate(bio)) {
-
-#if TLS_CHAIN_NO_SELFSIGNED // ignore self-signed certs in the chain
-            // self-signed certificates are not valid in a sent chain
-            if (SelfSigned(*ca)) {
-                debugs(83, DBG_PARSE_NOTE(2), "CA certificate is self-signed, will not be chained: " << *ca);
-                continue;
-            }
-#endif
             // checks that the chained certs are actually part of a chain for validating cert
             if (IssuedBy(*latestCert, *ca)) {
+
+                if (SelfSigned(*latestCert)) { // TODO: Rename to lastChained
+                    Assure(SelfSigned(*ca));
+                    Assure(!SelfSigned(*cert)); // TODO: Rename to leafCert or signingCert
+                    debugs(83, DBG_PARSE_NOTE(2), "WARNING: Ignoring repeated Root CA: " << *ca);
+                    continue;
+                }
+
                 debugs(83, DBG_PARSE_NOTE(3), "Adding issuer CA: " << *ca);
                 // OpenSSL API requires that we order certificates such that the
                 // chain can be appended directly into the on-wire traffic.
                 latestCert = CertPointer(ca);
                 chain.emplace_back(latestCert);
             } else {
+                // XXX: This usually logs a misleading "ignoring CA" message for
+                // the non-CA signing certificate which is not actually ignored!
```

Comment:
Your response implies that you disagree with something I wrote. Please answer the question posed:
> Did you mean this to say "non-CA server certificate" ?

Reply:
> > Removed [a comment marking an alleged out-of-scope problem] to avoid arguing.

> Your response implies that you disagree with something I wrote.

My response does not imply a disagreement (or agreement). There may be a way to find an interpretation of what you wrote that I can agree with, but that takes time. Making sure that you agree with that interpretation takes even more time. Adjusting the comment to reflect that consensus takes time. I just want to avoid spending more time on an out-of-scope comment, especially since the problem the comment describes is very likely to be resolved in future PRs (where spending time on getting to an agreement regarding the problem and the best solution would be appropriate).

In other words, I was hoping for a quick win -- an informal marker of a problem I knew about -- but was content with settling with the status quo -- official code with no markings. The suggested simple drop-it-or-leave-it procedure would have accomplished a satisfactory outcome without time waste. What we ended up having instead is the same no-markings outcome, but with a 20+ minutes waste (so far).

> Please answer the question posed: Did you mean this to say "non-CA server certificate" ?

The answer is "No, I did not", but that answer does not imply that your wording is wrong (i.e. has no interpretation that makes it correct). The (obviously imperfect) comment said what I meant it to say. The comment is gone now. The alleged problem the gone comment was marking is out of this PR scope. Can we cut our losses here and move on, please?

Reply:
Okay.